cvmfs/sync_union.cc

```diff
@@ -203,98 +281,36 @@ void SyncUnionOverlayfs::ProcessFile(SyncItem *entry) {
-                 scratch_path.c_str());
-        if (platform_lstat(scratch_path.c_str(), &scratch_stat) < 0) {
-          if (errno == ENOENT) {
-            // file is not present in scratch, warn, and/or abort
-            LogCvmfs(kLogUnionFs, kLogWarning,
-  "[WARNING] a file in the OverlayFS lowerdir (%s) has multiple\n"
-  "hard links, at least one of which (%s)\n"
-  "has been modified in CernVM-FS.  Due to a limitation in OverlayFS, after\n"
-  "this sync the modified file would no longer be part of its previous \n"
-  "hardlink group (CernVM-FS inode %"PRIu64").\n"
-  "The sync operation will now be aborted so that this issue can be corrected\n"
-  "manually -- you must explicitly include all files belonging to this\n"
-  "hardlink group in the CernVM-FS transaction (e.g. by touching them if you\n"
-  "wish to break the link or linking them if you wish to preserve it).\n"
-  "\n"
-  "To restore only this hardlink, you could run:\n"
-  "rm %s && ln %s %s\n"
-  "\n"
-  "To find all files that are part of this hardlink group, use:\n"
-  "find %s -inum %"PRIu64"\n"
-  "\n"
-  "To restore all hardlinks in this group, try something like:\n"
-  "for file in $(find %s -inum %"PRIu64"); do rm ${file} && ln %s ${file}; done"
-  "\n\n"
-  "Once you have corrected the issue, run `cvmfs_server publish` again\n",
-                     filename.c_str(),
-                     entry->GetUnionPath().c_str(),
-                     entry->GetRdOnlyInode(),
-                     union_path.c_str(),
-                     entry->GetUnionPath().c_str(), union_path.c_str(),
-                     union_parent_dir.c_str(), entry->GetRdOnlyInode(),
-                     union_parent_dir.c_str(), entry->GetRdOnlyInode(),
-                     entry->GetUnionPath().c_str());  // LogCvmfs
-            abort();
-          }  // error == ENOENT
-        }  // platform_lstat < 0
-      }  // for hardlink_lower_files_
-    }  // hardlink_lower_files_.size() == entry->GetRdOnlyLinkcount()
-  }  // Is hardlink
+  CheckForBrokenHardlink(entry);
+  MaskFileHardlinks(entry);
 
   SyncUnion::ProcessFile(entry);
 }
 
 
+void SyncUnionOverlayfs::CheckForBrokenHardlink(SyncItem *entry) const {
+  if (!entry->IsNew() && entry->GetRdOnlyLinkcount() > 1) {
+    LogCvmfs(kLogPublish, kLogStderr, "OverlayFS has copied-up a file (%s) "
+                                      "with existing hardlinks in lowerdir "
+                                      "(linkcount %d). OverlayFS cannot handle "
+                                      "hardlinks and would produce "
+                                      "inconsistencies. Aborting..." ,
+             entry->GetUnionPath().c_str(), entry->GetRdOnlyLinkcount());
+    abort();
+  }
+}
+
+void SyncUnionOverlayfs::MaskFileHardlinks(SyncItem *entry) const {
+  assert(entry->IsRegularFile() || entry->IsSymlink());
+  if (entry->GetUnionLinkcount() > 1) {
+    LogCvmfs(kLogPublish, kLogStderr, "Warning: Found file with linkcount > 1 "
```

Comment:
This might actually benefit from a silencing switch, since it might (and will) show up in production.